echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/postprocessors/BuildInfoPostProcessor.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2018 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.pipelinetriggers.postprocessors;
+
+import com.netflix.spinnaker.echo.model.Pipeline;
+import com.netflix.spinnaker.echo.model.Trigger;
+import com.netflix.spinnaker.echo.services.IgorService;
+import lombok.NonNull;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+import java.util.Map;
+
+/**
+ * Post-processor that looks up build details from Igor if a pipeline's trigger is associated
+ * with a build (ie, has a master and build number). Any later post-processor that wants to reference
+ * this build information should set its @Order to greater than 1.
+ */
+@Component
+@ConditionalOnProperty("igor.enabled")
+public class BuildInfoPostProcessor implements PipelinePostProcessor{
+  private IgorService igorService;
+
+  BuildInfoPostProcessor(@NonNull IgorService igorService) {
+    this.igorService = igorService;
+  }
+
+  public Pipeline processPipeline(Pipeline inputPipeline) {
+    Trigger inputTrigger = inputPipeline.getTrigger();
+    if (inputTrigger == null) {
+      return inputPipeline;
+    } else {
+      Trigger augmentedTrigger = addBuildInfo(inputTrigger);
+      return inputPipeline.withTrigger(augmentedTrigger);
+    }
+  }
+
+  private Trigger addBuildInfo(@NonNull Trigger inputTrigger) {
+    String master = inputTrigger.getMaster();
+    Integer buildNumber = inputTrigger.getBuildNumber();
+    String job = inputTrigger.getJob();
+    String propertyFile = inputTrigger.getPropertyFile();
+
+    Map<String, Object> buildInfo = null;
+    Map<String, Object> properties = null;
+    if (master != null && buildNumber != null && StringUtils.isNotEmpty(job)) {
+      buildInfo = igorService.getBuild(buildNumber, master, job);
```

Comment:
Do you know if `echo` using the `OkHttpMetricsInterceptor` from kork? At the very least, that would emit a metric on communication failures between echo and igor.

Reply:
I'd just copied this over from [Orca](https://github.com/spinnaker/orca/blob/a1e7eae28bcf3b443fdc6797a584e22fa829821e/orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy#L199) with the intention of removing this code from Orca once this merges.  The Orca code doesn't handle retries so I don't think this will be any worse---but it's a good point that it probably makes sense to add some retry logic.  I've made the calls to Igor try up to 5 times with exponential backoff in the event of failure.

Also, the configuration class we're using does inject the interceptor into the config ([here](https://github.com/spinnaker/kork/blob/58a5e6e4e146efea722fbcf2b6ab185141a74fae/kork-web/src/main/groovy/com/netflix/spinnaker/config/OkHttp3ClientConfiguration.groovy#L65)) so we should get metrics on communication failure.

I didn't go as far as to add Hystrix in to circuit-break this---it looks like the trend hasn't been to do this for intra-service communication (except for gate -> other services).  That being said, maybe it makes sense to do that?